### ISSUE TITLE
Compliance with Dask version 2021.03 ?

### ISSUE THREAD

**Issue by knutfrode** [xgcm/xhistogram#27]
First of all thank you for making this useful tool.
It has worked well for me until Dask version 2021.03, after which I am getting error as below when calling `histogram`. I apologize that this might be not be sufficient information to reproduce.

```
../../opendrift/models/basemodel.py:3707: in get_density_xarray
    h = histogram(self.ds.lon, self.ds.lat, bins=[lonbin, latbin],
../../../../miniconda3/envs/opendrift/lib/python3.9/site-packages/xhistogram/xarray.py:133: in histogram
    h_data = _histogram(*args_data, weights=weights_data, bins=bins, axis=axis,
../../../../miniconda3/envs/opendrift/lib/python3.9/site-packages/xhistogram/core.py:261: in histogram
    bin_counts = _histogram_2d_vectorized(*all_args_reshaped, bins=bins,
../../../../miniconda3/envs/opendrift/lib/python3.9/site-packages/xhistogram/core.py:133: in _histogram_2d_vectorized
    bin_counts = _dispatch_bincount(bin_indices, weights, N, hist_shapes,
../../../../miniconda3/envs/opendrift/lib/python3.9/site-packages/xhistogram/core.py:81: in _dispatch_bincount
    return _bincount_2d(bin_indices, weights, N, hist_shapes)
../../../../miniconda3/envs/opendrift/lib/python3.9/site-packages/xhistogram/core.py:32: in _bincount_2d
    return bc_offset.reshape(final_shape)
../../../../miniconda3/envs/opendrift/lib/python3.9/site-packages/dask/array/core.py:1919: in reshape
    return reshape(self, shape, merge_chunks=merge_chunks)
```

```
        if reduce(mul, shape, 1) != x.size:
>           raise ValueError("total size of new array must be unchanged")
E           ValueError: total size of new array must be unchanged

../../../../miniconda3/envs/opendrift/lib/python3.9/site-packages/dask/array/reshape.py:206: ValueError
```

**Comment by rabernat** [comment]
Ok, thanks for reporting this. Need to get a fix in asap. Any ideas?

Ideally we should be testing against xarray and dask master branches.

**Comment by raybellwaves** [comment]
Took a quick look at one of the failing tests in xskillscore https://github.com/xarray-contrib/xskillscore/issues/280#issuecomment-798825624

Failing at `dask/array/reshape.py:206`. Worth getting at with a minimal example.

**Comment by raybellwaves** [comment]
failing here with dask: https://github.com/raybellwaves/xhistogram/blob/master/xhistogram/test/test_core.py#L177

**Comment by rabernat** [comment]
The core issue is that dask's `bincount` function, which we import here

https://github.com/xgcm/xhistogram/blob/a3760d3fc1e0aca29f3a5180e7cbc1c5abc8f68f/xhistogram/duck_array_ops.py#L33

is now returning dask arrays with shape `()`. I'm looking into this. Is there a dask PR someone can point to where this behavior was changed?

**Comment by raybellwaves** [comment]
maybe https://github.com/dask/dask/pull/7391 will help?

**Comment by rabernat** [comment]
I just reviewed that PR and I'm 99% sure that it will fix the problem.

I consider this a regression in dask, so we will not be making a workaround in xhistogram. We have two choices here:
- Do nothing and wait until it's fixed + released in dask.
- Pin xhistogram to an earlier version of dask and do a bugfix release now.

What do those affected prefer?

**Comment by dougiesquire** [comment]
I'm fine with doing nothing if others are. dask releases are more or less monthly and I think we've already pinned an earlier version of dask in the last xskillscore release.

**Comment by jrbourbeau** [comment]
https://github.com/dask/dask/pull/7391 was just merged into Dask's `main` branch and I confirmed that tests here pass with the changes in that PR

**Comment by aaronspring** [comment]
I find a followup issue with `dask==2021.04`: https://github.com/xarray-contrib/xskillscore/issues/288

**Comment by raybellwaves** [comment]
xgcm tests seem ok on upstream: https://github.com/xgcm/xhistogram/actions/workflows/upstream.yml
Is there something in xskillscore that isn't tested in xgcm?

**Comment by aaronspring** [comment]
the xhist wrapper for datasets by @dougiesquire

**Comment by dougiesquire** [comment]
I think I've found the issue here. It's an issue in dask that was introduced into xhistogram in the latest version.

The dask version of `ravel_multi_index` ([introduced into xhistogram in 0.1.3](https://github.com/xgcm/xhistogram/pull/31)) does not actually replicate the API of the equivalent numpy function. Specifically, the [numpy implementation](https://numpy.org/doc/stable/reference/generated/numpy.ravel_multi_index.html) can receive a tuple of integer arrays, whereas the dask version cannot (despite saying it can). This functionality is required in xhistogram when there are multiple `args`. 

That is, the following works:
```python
import numpy as np
import dask.array as dsa

arr = np.array([[3,6,6],[4,5,1]]) # Example from numpy docs
da = dsa.from_array(arr)

dsa.ravel_multi_index(da, (7,7))
```
but the following fails with the error `'tuple' object has no attribute 'map_blocks'`:
```python
dsa.ravel_multi_index((da, da), (7,7))
```
This is a bug in dask and I'll open an issue about it.

xhistogram doesn't currently have tests with multiple dask `args`, which is why this didn't get flagged earlier (we should add them!), but xskillscore does try to use xhistogram in this way (and thus fails).

@raybellwaves, @aaronspring one solution for now in xskillscore is to pin xhistogram==0.1.2.

**Comment by dougiesquire** [comment]
Issue opened with dask here: https://github.com/dask/dask/issues/7580

**Comment by dougiesquire** [comment]
> Issue opened with dask here: [dask/dask#7580](https://github.com/dask/dask/issues/7580)

A fix for this has been implemented in the latest release of dask (2021.06.1). In the meantime, however, the xhistogram refactor to use `dask.array.blockwise` (#49) has removed our dependence on the dask.array functions referenced in this issue, so we actually no longer need this fix.

Closing the issue.
